src/events/upgrade.py

```diff
@@ -33,85 +38,108 @@ class KafkaDependencyModel(BaseModel):
     kafka_service: DependencyModel
 
 
-class KafkaUpgrade(DataUpgrade):
+class UpgradeManager(DataUpgrade):
     """Implementation of :class:`DataUpgrade` overrides for in-place upgrades."""
 
     def __init__(self, charm: "KafkaK8sCharm", **kwargs):
         super().__init__(charm, **kwargs)
         self.charm = charm
+        self.substrate = self.charm.state.substrate
 
-    @property
-    def current_version(self) -> str:
-        """Get current Kafka version."""
-        dependency_model: DependencyModel = getattr(self.dependency_model, "kafka_service")
-        return dependency_model.version
-
-    @property
-    def zookeeper_current_version(self) -> str:
-        """Get current Zookeeper version."""
-        return self.charm.state.zookeeper.get_zookeeper_version()
-
-    def post_upgrade_check(self) -> None:
-        """Runs necessary checks validating the unit is in a healthy state after upgrade."""
-        self.pre_upgrade_check()
-
-    @override
-    def pre_upgrade_check(self) -> None:
-        default_message = "Pre-upgrade check failed and cannot safely upgrade"
-        if not self.charm.healthy:
-            raise ClusterNotReadyError(message=default_message, cause="Cluster is not healthy")
+        self.framework.observe(
+            getattr(self.charm.on, "upgrade_charm"), self._on_kafka_pebble_ready_upgrade
+        )
 
-    @override
-    def build_upgrade_stack(self) -> list[int]:
-        upgrade_stack = []
-        units = set([self.charm.unit] + list(self.charm.peer_relation.units))  # type: ignore[reportOptionalMemberAccess]
-        for unit in units:
-            upgrade_stack.append(int(unit.name.split("/")[-1]))
-
-        return upgrade_stack
-
-    @override
-    def log_rollback_instructions(self) -> None:
-        logger.critical(ROLLBACK_INSTRUCTIONS)
+    def _on_kafka_pebble_ready_upgrade(self, event: EventBase) -> None:
```

Comment:
**question:** Why can't this hook follow the same pattern as vm? I mainly mean the call to `self.charm._on_update_status(event)` If that call is needed then my other comment about update_status check can be ignored.